### PR TITLE
Extend functionality to Find Command to search in specific Role Lists

### LINKED ISSUE #74: Support `find` command filtering separately for volunteers and befriendees

### PR DESCRIPTION
**Pull request by gerteck**
Support find command filtering separately for volunteers and befriendees.
Added testcases to test for find by role.

Current Format:
find KEYWORD [MORE_KEYWORDS]

Added feature: Filter separately for volunteer and befriendee, while not clearing the other list. This is to faciliate pairing and unpairing operations using the index.

`find r/volunteer KEYWORD [MORE_KEYWORDS]`
`find r/befriendee KEYWORD [MORE_KEYWORDS]`

Action: Will find given person in the list of the respective role, and leave other list untouched.

Closes #74

### CODE REVIEW

File: src/main/java/scrolls/elder/logic/commands/FindCommand.java [hunk 1]
```diff
@@ -9,26 +9,56 @@
...
+                            model.getFilteredVolunteerList().size(),
+                            "volunteer"));
+
+        } else if (searchBefriendee) {
+            model.updateFilteredBefriendeeList(predicate);
+            return new CommandResult(
+                    String.format(Messages.MESSAGE_PERSONS_LISTED_OVERVIEW_WITH_ROLE,
+                            model.getFilteredBefriendeeList().size(),
+                            "befriendee"));
+        }
```
Comment (CJerrong) [inline src/main/java/scrolls/elder/logic/commands/FindCommand.java]: Really like the avoidance of deep nesting here.

File: src/main/java/scrolls/elder/logic/parser/FindCommandParser.java [hunk 2]
```diff
@@ -25,9 +33,60 @@ public FindCommand parse(String args) throws ParseException {
...
+        } else {
+            return new FindCommand(new NameContainsKeywordsPredicate(Arrays.asList(newNameKeywords)),
+                    false, true);
+        }
+
+    }
+
+    private static FindType parseForRoles(String[] nameKeywords) throws ParseException {
+        boolean searchVolunteer = false;
+        boolean searchBefriendee = false;
```
Comment (CJerrong) [inline src/main/java/scrolls/elder/logic/parser/FindCommandParser.java]: Could perhaps rename these boolean variables to be more boolean sounding

File: src/main/java/scrolls/elder/model/Model.java [hunk 3]
```diff
@@ -89,4 +89,18 @@ public interface Model {
...
+     * @throws NullPointerException if {@code predicate} is null.
+     */
+    void updateFilteredVolunteerList(Predicate<Person> predicate);
+
+    /**
+     * Updates the filter of the filtered befriendee list to filter by the given {@code predicate}.
+     * Volunteer list is not filtered.
+     * @throws NullPointerException if {@code predicate} is null.
+     */
+    void updateFilteredBefriendeeList(Predicate<Person> predicate);
```
Comment (CJerrong) [inline src/main/java/scrolls/elder/model/Model.java]: Not sure if these commands have been phased out through jia en's latest pr, may have to refactor this.

File: src/main/java/scrolls/elder/model/ModelManager.java [hunk 4]
```diff
@@ -147,7 +147,25 @@ public void updateFilteredPersonList(Predicate<Person> predicate) {
...
+        filteredVolunteers.setPredicate(person -> predicate.test(person) && person.isVolunteer());
+    }
+
+    @Override
+    public void updateFilteredBefriendeeList(Predicate<Person> predicate) {
+        requireNonNull(predicate);
+
+        // Only apply predicate to befriendees
+        filteredPersons.setPredicate(person -> predicate.test(person) || person.isVolunteer());
+        filteredBefriendees.setPredicate(person -> predicate.test(person) && person.isBefriendee());
```
Comment (CJerrong) [inline src/main/java/scrolls/elder/model/ModelManager.java]: May have to move this functionality down into the Datastore structure

**Review by CJerrong — APPROVED** [review]
Overall, LGTM, may need some minor refactoring to be line with our latest design changes.

File: src/main/java/scrolls/elder/model/ModelManager.java (hunk 4 above)
Comment (gerteck) [reply src/main/java/scrolls/elder/model/ModelManager.java]: Done! Thank you for the insight!

File: src/main/java/scrolls/elder/logic/parser/FindCommandParser.java (hunk 2 above)
Comment (gerteck) [reply src/main/java/scrolls/elder/logic/parser/FindCommandParser.java]: Noted on this

**Pull request merged**